content/posts/2022-02-02-decreased-backend-tests-step/index.md

```diff
@@ -0,0 +1,173 @@
+
+CircleCI provides the option to split tests by timing, which was exactly what we were looking for. By doing so, we had our big moment: the test step decreased from about 22m to 6m.
+
+However, Coveralls started failing, since each job had a portion of coverage data. We had to combine multiple coverage files into a single one:
+
+```yaml
+    # ...
+    - run:
+    name: Run Python tests
+    command: |
+        TEST_FILES=$(circleci tests glob "head/**/test_*.py" | circleci tests split --split-by=timings)
+        pytest --no-migrations --junitxml=build/report_${CIRCLE_NODE_INDEX}.xml --cov-append --cov=head --cov-config .coveragerc --cov-report term --cov-report xml --capture=no -n 15 $TEST_FILES
+    - run: |
+        mv build/coverage.xml build/coverage_${CIRCLE_NODE_INDEX}.xml
+        mv .coverage build/.coverage_${CIRCLE_NODE_INDEX}
+    - store_test_results:
+        path: build
+    #...
+    - run:
+    name: Publish coverage
+    command: |
+        coverage combine build/.coverage_*
+        coveralls
+    #...
+```
+
+### Bonus action:
+
+CircleCI also provides a great feature to [cache the docker layer](https://circleci.com/docs/2.0/docker-layer-caching/). It may give you some more time improvements.
+In our case, we were already using it.
+
+## Lessons learned
+
+- Tech debts should not be ignored and fixes are WORTH the investment.
+
+In this particular large codebase, we had several tech debts that were found by this Working Group. Some of them we were able to fix, but this is something we as engineers should keep our eyes on, to avoid it existing for that long. As time goes by, it may become costly to pay those debts.
+
+- Test levels
+
+    We should be careful about different test levels. We should be aware of the [Test Pyramid](https://martinfowler.com/articles/practical-test-pyramid.html) to have a clear definition of the test boundaries. Should it mock external dependencies? Should we test things altogether (integration tests)?
+
+    Automated tests with external dependencies are slower than others. That is why it is really important to have a clear and well-defined Quality Strategy. What are the tools, practices, and processes we have to make sure that we are delivering good code?
+
+- Try to keep dependencies as updated as possible
+
+    Our software evolves as time goes by, such as libraries. Based on this, we should evaluate the tradeoffs of upgrading dependencies.
+
+
+By the end of this Working Group, we achieved our expected goals: backend tests step execution time running in about 6 minutes (73% of improvement).
+
+Even better than that, we had a great moment to reflect on our practices and raise improvement items.
+
+We have many open positions at the moment. Interested? Reach out or check out the [careers page](https://loadsmart.com/careers/).
```

Comment:
What about?

"Like to solve challenges like this one? We have many open positions at the moment. Check out our [engineering culture](https://github.com/loadsmart/culture) and the [careers page](https://loadsmart.com/careers/)."